### DISCUSSION TITLE
Is there any way to do basic maths or conditional loops?

### DISCUSSION THREAD

**Discussion by Extraltodeus** [comfyanonymous/ComfyUI#853] [Q&A]
Like loop for X times in the k-sampler while changing the seed every time, or have a node that would perform simple maths? Or a router that would route somewhere if condition else somewhere else? These feats would be more than interesting but I'm not finding them.

**Comment by GalaxyTimeMachine** [comment]
Perhaps this will help? https://github.com/pythongosssss/ComfyUI-Custom-Scripts/blob/main/py/math_expression.py
![image](https://github.com/comfyanonymous/ComfyUI/assets/52193044/432b3813-9a47-4d1d-8215-7d701a382534)

**Reply by Extraltodeus** [reply]
That is nice indeed! thank you! Still needs a way to loop however.

**Reply by ntdviet** [reply]
@GalaxyTimeMachine please share the name of the custom node you use to get these rectangular lines, thanks

**Reply by GalaxyTimeMachine** [reply]
Isn't this in your ComfyUI settings? I thought it had been added there now.

**Reply by ntdviet** [reply]
@GalaxyTimeMachine  oh shoot, I didn't catch up with the update, since when?...
Thanks btw

**Comment by Extraltodeus** [comment]
I ended up doing a lot of custom nodes in the end, including a way to save conditioning which can be then use as some sort of text embedding, also two clip interrogator nodes :)

https://github.com/Extraltodeus/CustomComfyUINodes

**Reply by ltdrdata** [reply]
Can the configuration of the repo be organized into a form that can be managed at once by creating `__init__.py`? In the current setup, since it is configured as it is, updates and other support cannot be provided as individual file units from the ComfyUI-Manager.

**Reply by Extraltodeus** [reply]
The ComfyUI-Manager?
I've dwelved into the nodes but I don't really know what you are talking about.

**Comment by ltdrdata** [comment]
> The ComfyUI-Manager?
> I've dwelved into the nodes but I don't really know what you are talking about.

https://github.com/ltdrdata/ComfyUI-Manager

This is the ComfyUI-Manager.
I am updating the DB periodically.

**Reply by Extraltodeus** [reply]
Pretty nice!
What do you need me to do? I wouldn't recommand to add each and every of my nodes tho. The most interesting might be the clip interrogator. The oobabooba is an adapted copy of another one so to make it work with the same name webui for the llama models for example so I don't know if it would be appreciated globally since it's relatively not from me. The rest of it is more or less a treasure cove. There are good and half-assed things.

**Reply by ntdviet** [reply]
You can reorganize your work into another repo that's ready to be published. Great work btw

**Reply by Extraltodeus** [reply]
Thank you :)

I'm not sharing everything since I'm mostly doing messy code while experimenting and I don't think that everybody would want my mess as a base to generate anything but as I am getting better I will share more interesting things. You can check my perlin-based noise generator for example. It makes it possible to get similar noise patterns at different scales :)
https://github.com/Extraltodeus/noise_latent_perlinpinpin

**Reply by ntdviet** [reply]
@Extraltodeus thanks for the link, but you give me a lot to process :')) Can you elaborate more on...
1. What's the practical use for this perlin noise? I mean, from a visual designer's angle, what's the difference compared to the usual gausian noise? Just another variation to the final image? Or improved quality (sharpness, dynamic range...)?
2. Do I understand your example workflow correctly, that it demonstrates how the noise you "artificially" created on the side can be used to seamlessly move from base to the refiner's sampling, as if the refiner sampler was dealing with leftover noise from the base sampler?
3. If my understanding of 2 is correct, and you also showed the possibility of upscaled perlin-noisy latent, can I do something like this: base sampling from step 1 - N > upscale the noisy latent > refiner sampling from step N - final step
4. If 3 is possible, I could interpose the noisy latent from SDXL1 to SD1.5, and finish the refining with another SD1.5 model, right? That would be frkin awesome...

**Reply by Extraltodeus** [reply]
1. Well according to a very little batch of testers my workflow produces sharper images. This might be also due to the time I've spent trying to figure out the best ratios to use with the upscaler and noise injection. The ultimate goal of my experiment is to be able to upscale in between the base and the refiner or somewhere along the generation process by reinjecting the SAME noise pattern but at a different scale. Since stable diffusion uses nosie to create images, reusing the same noise pattern should allow to upscale while keeping a more coherent result.
2. The workflow is more about just showing that the noise is viable, opening the way for the idea that I descripted above.
3. Yes you do understand it correctly and there is an example on the readme. The rabbit has been upscaled at 1.5x the resolution. Now I just managed to get the whole thing working so I would not be able to provide a completely working workflow to do bigger upscalings. That's kind of a proof of concept or "base that you can digg". Even tho I did eventually manage to do x2 upscales I am uncertain of my methods and do not deem worth sharing. As it is right now.
4. You can try to combine 1.5 with SDXL but the latent spaces are different. You should absolutely check [city96](https://github.com/city96/SD-Latent-Interposer) latent interposer if you want to attempt such thing.

I did manage yesterday to generate images from 1.5 to the SDXL refiner but it was really finicky and I didn't used the latent interposer solution as I was only experimenting.

**Reply by ntdviet** [reply]
@Extraltodeus  THANKS for quick reply. That's fantastic news. I had been working on a workflow to blend SD1.5 in to SDXL1 base image and wasn't able to work on the latent plane cause latent up-scaling corrupts the leftover noise. Had to switch back and forth to pixel a lot, which is pretty time and memory consuming at higher resolution. Now your solution will help me get back on track with direct latent & noise manipulation. Thanks a bunch!!!
Btw, this is my workflow as of now, before I'm using your input to modify: https://github.com/ntdviet/comfyui-ext/tree/main/custom_workflows/SDXL1.0_SD1.5_Mix_FixTune Maybe you can have a look and give me some improvement ideas, thanks.

Regarding city96's latent interposer, I had it on my workflow before, but the quality was not satisfying. Might try again with perlin noise latents, hopefully, it's better this time

**Reply by Extraltodeus** [reply]
No problem! If you manage to get good results I am very curious!

**Reply by ltdrdata** [reply]
If you're experimenting with latent, I recommend checking out this extension as well. It offers latent upscale based on neural network to minimize damage.

https://github.com/city96/SD-Latent-Upscaler

**Reply by Extraltodeus** [reply]
@ltdrdata haha yes he made it after I requested his help.

https://github.com/city96/SD-Advanced-Noise/issues/1

I just got my first good results

![00073UI_00001_](https://github.com/comfyanonymous/ComfyUI/assets/15731540/bd5eb1b9-b2b9-4c23-8573-13bdc10e2e61)

edit: @ntdviet check this out :)

**Reply by ntdviet** [reply]
HOLY F...!!! I was about to complain that latent upscale is still messed up after the noise injection, so it was definitely not because of leftover noise missing :)) Need to try this out asap. If I got it right from your chat with city96, I might even not need to do noise injection but his model can handle leftover noise very well, right?

Btw, I tried Perlin noise injection without upscaling the latent, just the noise upscaled from 1024 to 1536. And it kind of airbrushes the ultra fine details to directional lines. The small feathers here on the left got combed :))  
![ảnh](https://github.com/comfyanonymous/ComfyUI/assets/54492570/a6f82482-fbd0-437c-99f3-e9209def6a53)
Changing the noise generator's parameters didn't solve it, but when I do overlapped sampling by about 10% of total steps, the details could be preserved.

**Reply by Extraltodeus** [reply]
I never use leftover noise but rather always reinject the right amount before sampling.

I shared a [node](https://civitai.com/models/121394) on civitai that automates this idea.

> I tried Perlin noise injection without upscaling the latent, just the noise upscaled from 1024 to 1536. 

The noise injector does not inject anything if the noise latent is not at the same shape. Check your terminal if it is not warning you about it. If you try one of the workflows included in the shared image in the discussion with city96 you should probably be able to get an even higher quality.

**Reply by Extraltodeus** [reply]
@ntdviet this is what I mean:

![00162UI_00001_](https://github.com/comfyanonymous/ComfyUI/assets/15731540/ce59f1cb-f1d7-42ad-931c-996e6e952eba)

![00164UI_00001_](https://github.com/comfyanonymous/ComfyUI/assets/15731540/488a07b4-fbc7-4b8f-8fc4-a85c17a5186e)

The level of detail is ridiculous

**Reply by Extraltodeus** [reply]
For a fair comparison I injected usual noise for the upscale, keeping the perlin for the base so to keep the same base image, and it does indeed not give the same result:

![00216UI_00001_](https://github.com/comfyanonymous/ComfyUI/assets/15731540/c56fd782-6e24-49c6-946d-53c199d7272d)

**Reply by city96** [reply]
Figured I'd add my 2¢ here as well.

@ntdviet 

>Regarding city96's latent interposer, I had it on my workflow before, but the quality was not satisfying.

Fair, I was only using it to get a rough composition using SDXL and then using it at a very high denoise. Both the interposer and the upscaler are pretty bad in terms of actual quality since I have zero experience in any of this ML stuff. I have some ideas to improve them but alas there's only so many hours in a day.

@Extraltodeus 

I've tested the new workflow and it seems to work well enough, even on anime. I can't test extensively since I'm not at my PC, but some things I noticed:

- `Get sigma` needs to be connected to the base model when using a LoRA. Connecting it to the LoRA model output introduces too much noise in stage 2.

- Using two different models for the first/second stage behaved weirdly for me. Might also be connected to the sigma stuff.

- There is a noticeable lack of reproducibility/determinism for me. With previews enabled, canceling the execution mid-generation (KSampler running) and starting it again without changing anything results in a different image.


Also, I think it would be best to start a new discussion topic here on the main ComfyUI repo related to all the noise experiments. That way we can collect everything centrally instead of having it spread out over multiple issues/discussions/repos.


![ComfyUI_01185_](https://github.com/comfyanonymous/ComfyUI/assets/125218114/fedb1b39-f8e0-44dd-9b39-a1899c4679b1)

**Reply by ntdviet** [reply]
> The noise injector does not inject anything if the noise latent is not at the same shape. Check your terminal if it is not warning you about it. If you try one of the workflows included in the shared image in the discussion with city96 you should probably be able to get an even higher quality.

Ah ok, so I can not inject 1056x1056 noise into 1024x1024 latents? good to know thanks. It didn't throw any warning for me. Thought I could cheat for more details by having a higher resolution noise map for smaller latents.

I used your guy's workflows to compare perlin noise with regular leftover noise and it was pretty disappointing for my SD1.5 checkpoint, totally blurred. The version with leftover noise got it pretty decent, though not perfect as you already stated.
![ComfyUI_temp_jplsg_00006_](https://github.com/comfyanonymous/ComfyUI/assets/54492570/f76e8c83-f6d2-4c49-be6a-69957fef8d42)
![ảnh](https://github.com/comfyanonymous/ComfyUI/assets/54492570/3aa46df4-f2c7-4a60-9c71-5ddb023ef4ce)

Then I got surprised by the model pic, and yes, for model closeups, perlin noise performs pretty well.
In this kind of extreme pics, it outperforms normal noise by a lot :))
![ComfyUI_temp_jplsg_00008_](https://github.com/comfyanonymous/ComfyUI/assets/54492570/71b83514-ac86-4639-afd3-55438f51839f)
The hires-fix of faces is clearly better and more consistent with perlin
![ảnh](https://github.com/comfyanonymous/ComfyUI/assets/54492570/eac854f0-0733-45db-aac1-6afe107a8628)

So why is it shit for buildings but good for characters? Because of the checkpoint I used? Can I have yours? Couldn't find the ckpt you used in your workflow

**Reply by Extraltodeus** [reply]
@city96 In my experience you always have to use the model used to generate the image to get the right sigma.

>Also, I think it would be best to start a new discussion topic here on the main ComfyUI repo related to all the noise experiments. That way we can collect everything centrally instead of having it spread out over multiple issues/discussions/repos.

Agreed! The issue open on your repo might be a good place unless you have other thoughts.

**Reply by Extraltodeus** [reply]
@ntdviet but here again you are doing the same mistake of not using the same resolution! That's why you result is blurry. This is because your second pass does not actually have noise.

![image](https://github.com/comfyanonymous/ComfyUI/assets/15731540/a9e9e27a-49e8-4e1a-ab5f-bcc7b6cfad73)

**Reply by ntdviet** [reply]
@Extraltodeus oh damn, ok, sorry for the error, I thought I checked everything. But then why is it working for the girls :))
Edit: sorry I'm the idiot, I got the upscale correctly for the girls, x2

**Reply by city96** [reply]
@Extraltodeus 

My personal vote is for this current repo (the main ComfyUI one), since that's the most likely to get traction/other people joining in. 

The issue on my repo could work, but I prefer discussions since you can have multiple top level comments at the same time for different things.

**Comment by ltdrdata** [comment]
> Pretty nice! What do you need me to do? I wouldn't recommand to add each and every of my nodes tho. The most interesting might be the clip interrogator. The oobabooba is an adapted copy of another one so to make it work with the same name webui for the llama models for example so I don't know if it would be appreciated globally since it's relatively not from me. The rest of it is more or less a treasure cove. There are good and half-assed things.

This depends on the developer of the node. They can choose to provide useful nodes and miscellaneous nodes separately in a separate pack, or they can offer options in the configuration to enable or disable certain functionalities.

In my case, I provide three separate extensions based on the nature of the nodes. Users can receive the specific set of nodes/extension they desire.

On the other hand, for legacy nodes with messy dependencies, I offer them in a disabled state, and provide users with a configuration option to selectively enable them if they choose to.

You can choose how you want to provide it to the users, depending on your preference. However, I would suggest offering it in a form that can be easily installed using `git clone`, including `__init__.py`. This way, it can be easily tracked for updates when you make changes in the future.

**Reply by Extraltodeus** [reply]
I've done one that is worth sharing, I've included the init file too!
https://github.com/Extraltodeus/noise_latent_perlinpinpin
